File: 1256-rank-transform-of-an-array/1256-rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        vector<int> sorted = arr;
        sort(sorted.begin(), sorted.end());
        
        unordered_map<int, int> rank;
        int r = 1;
        
        for (int num : sorted) {
            if (rank.count(num) == 0) {
                rank[num] = r++;
            }
        }

        vector<int> result;
        for (int num : arr) {
            result.push_back(rank[num]);
        }

        return result;
    }
};

File: 1256-rank-transform-of-an-array/1256-rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        int[] copy = arr.clone();
        Arrays.sort(copy);

        Map<Integer, Integer> map = new HashMap<>();
        int rank = 1;

        for (int num : copy) {
            if (!map.containsKey(num)) {
                map.put(num, rank++);
            }
        }

        int[] result = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            result[i] = map.get(arr[i]);
        }

        return result;
    }
}
